File: public/dsa/arrays/merge_two_sorted_arrays/optimal/merge_two_sorted_arrays.cpp

#include <bits/stdc++.h>
using namespace std;

vector<int> mergeSortedArrays(vector<int> A, vector<int> B) {
    vector<int> C(A.size() + B.size());
    int i = 0, j = 0, k = 0;
    while (i < A.size() && j < B.size()) {
        if (A[i] < B[j]) {
            C[k] = A[i];
            k++;
            i++;
        } else {
            C[k] = B[j];
            k++;
            j++;
        }
    }
    while (i < A.size()) {
        C[k] = A[i];
        k++;
        i++;
    }
    while (j < B.size()) {
        C[k] = B[j];
        k++;
        j++;
    }
    return C;
}

// tests-start
int main() {
    vector<int> A = {1, 2, 3, 4, 4};
    vector<int> B = {2, 4, 5, 5};
    vector<int> C = mergeSortedArrays(A, B);
    for (int v : C) {
        cout << v << " ";
    }
    cout << endl;
    return 0;
}
// tests-end

/*output
1 2 2 3 4 4 4 5 5
*/

File: public/dsa/arrays/merge_two_sorted_arrays/optimal/merge_two_sorted_arrays.py
from typing import List


class Solution:
    def mergeSortedArrays(self, A: List[int], B: List[int]) -> List[int]:
        nA = len(A)
        nB = len(B)
        nC = nA + nB
        C = [0 for _ in range(nC)]
        iA = 0
        iB = 0
        iC = 0
        while iA < nA and iB < nB:
            if A[iA] < B[iB]:
                C[iC] = A[iA]
                iA += 1
                iC += 1
            else:
                C[iC] = B[iB]
                iB += 1
                iC += 1
        while iA < nA:
            C[iC] = A[iA]
            iA += 1
            iC += 1
        while iB < nB:
            C[iC] = B[iB]
            iB += 1
            iC += 1
        return C


# tests-start
A = [1, 2, 3, 4, 4]
B = [2, 4, 5, 5]
sol = Solution()
print(sol.mergeSortedArrays(A, B))
# tests-end

'''output
[1, 2, 2, 3, 4, 4, 4, 5, 5]
'''
